lib/SPIRV/SPIRVReader.cpp

```diff
@@ -3542,7 +3325,14 @@ bool SPIRVToLLVM::translate() {
   if (!transSourceExtension())
     return false;
   transGeneratorMD();
-  if (!transOCLBuiltinsFromVariables())
+  // TODO: add an option to control the builtin format in SPV-IR.
+  // The primary format should be function calls:
+  //   e.g. call spir_func i32 @_Z29__spirv_BuiltInGlobalLinearIdv()
+  // The secondary format should be global variables:
+  //   e.g. load i32, i32* @__spirv_BuiltInGlobalLinearId, align 4
+  // If the desired format is global variables, we don't have to lower them
+  // as calls.
+  if (!lowerBuiltinVariablesToCalls(M))
```

Comment:
While we don't have an option to control output format, do we need another call to `lowerBuiltinVariablesToCalls` from SPIRToOCL passes?

Reply:
Yes, I understand that, thanks.
I see that `lowerBuiltinVariablesToCalls` is also called at the end of SPIRVReader, I guess it already makes SPIRVReader to produce calls to work item built-ins (instead of variables use). After that all that is left to do in SPIRVToOCL passes is to change built-in names in visitCallSPIRVBuiltin.
But `lowerBuiltinVariablesToCalls` is called again at the beginning of SPIRVToOCL pass. Do we really need this second call, or I just misunderstood something? (Sorry If I did).

Reply:
I guess a comment on line 3328 "// TODO: add an option to control the builtin format in SPV-IR." explains why `lowerBuiltinVariablesToCalls` is being called twice and foresees a time when it actually won't be a case . BTW, why do we need to call `lowerBuiltinVariablesToCalls` in the reader in the first place? I'd prefer to always have GV-like built-ins in SPV-IR, or I just don't have enough data?

Reply:
> I guess a comment on line 3328 "// TODO: add an option to control the builtin format in SPV-IR." explains why lowerBuiltinVariablesToCalls is being called twice and foresees a time when it actually won't be a case . 
Yes, as I already mentioned in my first comment, while we don't have an option, can we make only one call?

> BTW, why do we need to call lowerBuiltinVariablesToCalls in the reader in the first place? I'd prefer to always have GV-like built-ins in SPV-IR
The thing is, we already have users that use calls, so we don't want to break their workflow.

Reply:
> Yes, as I already mentioned in my first comment, while we don't have an option, can we make only one call?

Sorry I didn't make it clear. The thing is, we have users that use the `SPIRVToOCL` pass independently, translating SPV-IR into OCL-IR. And for these users, the `SPIRVToOCL` pass may receive SPV-IR of both formats (builtin calls / variables) as input. For example, clang is still emitting builtin-variable-based SPV-IR.
So I think this call in `SPIRVToOCL` is probably unavoidable, at least for the current stage :(

Reply:
Ah, I understand. Thank you.